File: hooks/ponytail/ponytail-activate.js
#!/usr/bin/env node
// ponytail — Claude Code SessionStart activation hook
//
// Runs on every session start:
//   1. Writes flag file at $CLAUDE_CONFIG_DIR/.ponytail-active (defaults to ~/.claude; statusline reads this)
//   2. Emits ponytail ruleset as hidden SessionStart context
//   3. Detects missing statusline config and emits setup nudge

const fs = require('fs');
const path = require('path');
const { getDefaultMode, getClaudeDir, isShellSafe } = require('./ponytail-config');
const { getPonytailInstructions } = require('./ponytail-instructions');
const {
  clearMode,
  setMode,
  writeHookOutput,
} = require('./ponytail-runtime');

const claudeDir = getClaudeDir();
const settingsPath = path.join(claudeDir, 'settings.json');

const mode = getDefaultMode();

// "off" mode — skip activation entirely, don't write flag or emit rules
if (mode === 'off') {
  clearMode();
  writeHookOutput('SessionStart', 'OK');
  process.exit(0);
}

// 1. Write flag file
try {
  setMode(mode);
} catch (e) {
  // Silent fail -- flag is best-effort, don't block the hook
}

// 2. Emit the ponytail ruleset, filtered to the active intensity level.
let output = getPonytailInstructions(mode);

// 3. Detect missing statusline config — nudge Claude to help set it up
try {
  let hasStatusline = false;
  if (fs.existsSync(settingsPath)) {
    // Strip UTF-8 BOM some editors prepend on Windows (breaks JSON.parse)
    const raw = fs.readFileSync(settingsPath, 'utf8').replace(/^\uFEFF/, '');
    const settings = JSON.parse(raw);
    if (settings.statusLine) {
      hasStatusline = true;
    }
  }

  // Nudge at most once — the flag file marks that the user has already seen
  // (and implicitly declined) the statusline setup offer. Repeating it every
  // session start turns a helpful hint into a nag.
  const nudgeFlagPath = path.join(claudeDir, '.ponytail-statusline-nudged');
  if (!hasStatusline && !fs.existsSync(nudgeFlagPath)) {
    try { fs.writeFileSync(nudgeFlagPath, ''); } catch (e) { /* best-effort */ }
    const scriptPath = path.join(claudeDir, 'statusline_new.sh');
    if (isShellSafe(scriptPath)) {
      const command = `bash "${scriptPath}"`;
      const statusLineSnippet =
        '"statusLine": { "type": "command", "command": ' + JSON.stringify(command) + ' }';
      output += "\n\n" +
        "CONFIGURAZIONE STATUSLINE NECESSARIA: il plugin ponytail include un badge statusline che mostra la modalità attiva " +
        "(es. [PONYTAIL]). Non è ancora configurato. " +
        "Per abilitarlo, aggiungi questo a " + settingsPath + ": " +
        statusLineSnippet + " " +
        "Proponi proattivamente di configurarlo per l'utente alla prima interazione.";
    } else {
      // ponytail: il path di installazione contiene metacaratteri di shell — non includerlo in uno
      // snippet di comando; fai in modo che l'agente lo configuri a mano.
      output += "\n\n" +
        "CONFIGURAZIONE STATUSLINE NECESSARIA: il plugin ponytail include un badge statusline che mostra la modalità attiva. " +
        "Il suo path di installazione contiene caratteri non sicuri da includere in un comando shell, quindi configuralo manualmente: " +
        "aggiungi un comando statusLine di tipo \"command\" che esegue " + scriptPath +
        " in " + settingsPath + ", citando/effettuando l'escape del path per la tua shell. " +
        "Proponi proattivamente di configurarlo per l'utente alla prima interazione.";
    }
  }
} catch (e) {
  // Silent fail — don't block session start over statusline detection
}

try {
  writeHookOutput('SessionStart', output);
} catch (e) {
  // Silent fail — stdout closed/EPIPE at hook exit must not surface as a hook failure
}

File: hooks/ponytail/ponytail-activate.lua
#!/usr/bin/env lua
-- ponytail — Claude Code SessionStart activation hook. Versione Lua.

local script_dir = (arg[0] or "."):match("^(.*)[/\\]") or "."
local json = dofile(script_dir .. "/../lib/dkjson.lua")
local config = dofile(script_dir .. "/ponytail-config.lua")
local instructions = dofile(script_dir .. "/ponytail-instructions.lua")

local claude_dir = config.get_claude_dir()
local settings_path = claude_dir .. "/settings.json"
local nudge_flag = claude_dir .. "/.ponytail-statusline-nudged"

local mode = config.get_default_mode()
if mode == "off" then
  config.clear_mode()
  io.stdout:write("OK")
  os.exit(0)
end

config.set_mode(mode)
local output = instructions.get_instructions(mode)

-- Nudge statusline mancante, al massimo una volta (flag file).
local has_statusline = false
local raw = config.file_read(settings_path)
if raw then
  raw = raw:gsub("^\239\187\191", "") -- strip UTF-8 BOM
  local ok, settings = pcall(json.decode, raw)
  if ok and type(settings) == "table" and settings.statusLine then
    has_statusline = true
  end
end

if not has_statusline and not config.file_exists(nudge_flag) then
  config.file_write(nudge_flag, "")
  local script_path = claude_dir .. "/statusline_new.sh"
  if config.is_shell_safe(script_path) then
    local command = 'bash "' .. script_path .. '"'
    local snippet = '"statusLine": { "type": "command", "command": ' .. json.encode(command) .. ' }'
    output = output .. "\n\n" ..
      "CONFIGURAZIONE STATUSLINE NECESSARIA: il plugin ponytail include un badge statusline che mostra la modalità attiva " ..
      "(es. [PONYTAIL]). Non è ancora configurato. " ..
      "Per abilitarlo, aggiungi questo a " .. settings_path .. ": " ..
      snippet .. " " ..
      "Proponi proattivamente di configurarlo per l'utente alla prima interazione."
  else
    output = output .. "\n\n" ..
      "CONFIGURAZIONE STATUSLINE NECESSARIA: il plugin ponytail include un badge statusline che mostra la modalità attiva. " ..
      "Il suo path di installazione contiene caratteri non sicuri da includere in un comando shell, quindi configuralo manualmente: " ..
      'aggiungi un comando statusLine di tipo "command" che esegue ' .. script_path ..
      " in " .. settings_path .. ", citando/effettuando l'escape del path per la tua shell. " ..
      "Proponi proattivamente di configurarlo per l'utente alla prima interazione."
  end
end

io.stdout:write(output)
